src/main/java/org/truffleruby/language/yield/YieldExpressionNode.java

```diff
@@ -51,13 +53,24 @@ public final Object execute(VirtualFrame frame) {
             useCapturedBlock.enter();
 
             block = RubyArguments.getMethod(frame).getCapturedBlock();
+        }
+
+        if (block == null) {
+            // TODO (pitr-ch 19-Dec-2017): a hack to get zlib working, needs proper investigation
+            // captured block passing down is messy, it abuses blockArgument which is suppose to be left null
+            useDeclarationFrame.enter();
 
-            if (block == null) {
-                noCapturedBlock.enter();
-                throw new RaiseException(coreExceptions().noBlockToYieldTo(this));
+            final MaterializedFrame declarationFrame = RubyArguments.getDeclarationFrame(frame);
+            if (declarationFrame != null) {
+                block = RubyArguments.getBlock(declarationFrame);
             }
```

Comment:
Which piece code is needing this specifically?

Reply:
Sorry I don't remember, You'll need to comment out and run zlib tests.
I'll be looking at a proper fix, block passing combined with C functions seems incorrect.

Reply:
Can you find out and point to it here?
I think it is necessary to know where the hack come from to maintain this piece of code (asking people reading this comment to find out themselves is not efficient).

Reply:
This is a **hack** and I'll be looking for a proper fix soon. I don't want to loose time now looking up what the failing tests were, unfortunately I did not note it down. I'll find out when I go back to fix it.

Reply:
I disagree, please find where is the problem from in this PR and note it in the comment.
Hacks ought to at least point to the relevant problem.

Reply:
Please add a comment explaining what you know about why this check is needed and try to remove in a subsequent PR.